src/helpers/survey.js

```diff
@@ -0,0 +1,239 @@
+  setSurveyAudioPlayerBuffers(parsedLocale, bufferList);
+  setSurveyAudioLoading(false);
+}
+
+// Function to fetch buffer or return from the cache
+export const fetchBuffer = ({
+  parsedLocale,
+  setSurveyAudioLoading,
+  audioLinks,
+  surveyAudioBuffers,
+  setSurveyAudioPlayerBuffers,
+}) => {
+  const context = new AudioContext();
+  // buffer already exists for the given local
+  if (surveyAudioBuffers[parsedLocale]) {
+    return;
+  }
+  setSurveyAudioLoading(true);
+  const bufferLoader = new BufferLoader(context, audioLinks[parsedLocale], (bufferList) =>
+    finishedLoading({ bufferList, parsedLocale, setSurveyAudioLoading, setSurveyAudioPlayerBuffers }),
+  );
+
+  bufferLoader.load();
+};
+
+export const showAndPlaceAudioButton = ({ playAudioButton, el }) => {
+  if (playAudioButton) {
+    playAudioButton.classList.add('play-button-visible');
+    playAudioButton.style.display = 'flex';
+    el.appendChild(playAudioButton);
+  }
+};
+
+export function restoreSurveyData({ surveyInstance, uid, selectedAdmin, surveyResponsesData, surveyStore }) {
+  // Try to get data from localStorage first
+  const prevData = window.localStorage.getItem(`${LEVANTE_SURVEY_RESPONSES_KEY}-${uid}`);
+  if (prevData) {
+    const parsedData = JSON.parse(prevData);
+    surveyInstance.data = parsedData.responses;
+    surveyInstance.currentPageNo = parsedData.pageNo;
+    return { isRestored: true, pageNo: parsedData.pageNo };
+  } else if (surveyResponsesData) {
+    // If not in localStorage, try to find data from the server
+    const surveyResponse = surveyResponsesData.find((doc) => doc?.administrationId === selectedAdmin);
+    if (surveyResponse) {
+      if (!surveyStore.isGeneralSurveyComplete) {
+        surveyInstance.data = surveyResponse.general.responses;
+      } else {
+        const specificIndex = surveyStore.specificSurveyRelationIndex;
+        surveyInstance.data = surveyResponse.specific[specificIndex].responses;
+      }
+
+      surveyInstance.currentPageNo = surveyResponse.pageNo;
+      return { isRestored: true, pageNo: surveyResponse.pageNo };
+    }
+  }
+
+  // If there's no data in localStorage and no data from the server,
+  // the survey has never been started, so we continue with an empty survey
+  return { isRestored: false, pageNo: 0 };
+}
+
+export function saveSurveyData({
+  survey,
+  roarfirekit,
+  uid,
+  selectedAdmin,
+  questionName,
+  responseValue,
+  specificIds,
+  userType,
+  surveyStore,
+}) {
+  const currentPageNo = survey.currentPageNo;
+
+  if (window.localStorage.getItem(`${LEVANTE_SURVEY_RESPONSES_KEY}-${uid}`)) {
+    const prevData = JSON.parse(window.localStorage.getItem(`${LEVANTE_SURVEY_RESPONSES_KEY}-${uid}`));
+
+    // Update the page number at the top level
+    prevData.pageNo = currentPageNo;
+    prevData.responses[questionName] = responseValue;
+
+    window.localStorage.setItem(`${LEVANTE_SURVEY_RESPONSES_KEY}-${uid}`, JSON.stringify(prevData));
```

Comment:
Not a big deal since this is just reading from localStorage and not making any requests, but: 

```suggestion
  const surveyResponsesData = getSurveyResponsesData()
  
  if (surveyResponsesData) {
    // Update the page number at the top level
    surveyResponsesData.pageNo = currentPageNo;
    surveyResponsesData.responses[questionName] = responseValue;

    setSurveyResponseData(userId, surveyResponsesData)
```